controllers/trainingjob/trainingjob_controller_test.go

```diff
@@ -617,6 +617,294 @@ var _ = Describe("Reconciling a TrainingJob that exists", func() {
+				describeOutput := CreateDescribeOutputWithDebugJobStatus(expectedStatus, expectedSecondaryStatus, debugJobStatus)
+				describeOutput.FailureReason = ToStringPtr(failureReason)
+
+				mockSageMakerClientBuilder.
+					AddDescribeTrainingJobResponse(describeOutput)
+			})
+
+			When("!HasDeletionTimestamp", func() {
+				It("Requeues after interval", func() {
+					ExpectRequeueAfterInterval(reconcileResult, reconcileError, pollDuration)
+				})
+
+				It("Updates status", func() {
+					ExpectStatusToBe(trainingJob, string(expectedStatus), string(expectedSecondaryStatus))
+				})
+
+				It("Has the additional field set", func() {
+					ExpectAdditionalToContain(trainingJob, failureReason)
+				})
+
+				Context("Does not have a finalizer", func() {
+					BeforeEach(func() {
+						shouldHaveFinalizer = false
+					})
+
+					It("Adds a finalizer", func() {
+						ExpectToHaveFinalizer(trainingJob, controllers.SageMakerResourceFinalizerName)
+					})
+				})
+			})
+
+			When("HasDeletionTimestamp", func() {
+				BeforeEach(func() {
+					shouldHaveDeletionTimestamp = true
+				})
+
+				It("Deletes the training job", func() {
+					ExpectTrainingJobToBeDeleted(trainingJob)
+				})
+
+				It("Requeues after interval", func() {
+					ExpectRequeueAfterInterval(reconcileResult, reconcileError, pollDuration)
+				})
+			})
+		})
+
+		Context("TrainingJob with debug job is `stopping` has status 'Failed'", func() {
+			var failureReason string
+			var debugJobStatus sagemaker.RuleEvaluationStatus
+
+			BeforeEach(func() {
+				expectedStatus = sagemaker.TrainingJobStatusFailed
+				failureReason = "Failure within the training job"
+				debugJobStatus = sagemaker.RuleEvaluationStatusStopping
+
+				// Add the failure reason to the describe output
+				describeOutput := CreateDescribeOutputWithDebugJobStatus(expectedStatus, expectedSecondaryStatus, debugJobStatus)
+				describeOutput.FailureReason = ToStringPtr(failureReason)
+
+				mockSageMakerClientBuilder.
+					AddDescribeTrainingJobResponse(describeOutput)
+			})
+
+			When("!HasDeletionTimestamp", func() {
+				It("Requeues after interval", func() {
+					ExpectRequeueAfterInterval(reconcileResult, reconcileError, pollDuration)
+				})
+
+				It("Updates status", func() {
+					ExpectStatusToBe(trainingJob, string(expectedStatus), string(expectedSecondaryStatus))
+				})
+
+				It("Has the additional field set", func() {
+					ExpectAdditionalToContain(trainingJob, failureReason)
+				})
+
+				Context("Does not have a finalizer", func() {
+					BeforeEach(func() {
+						shouldHaveFinalizer = false
+					})
+
+					It("Adds a finalizer", func() {
+						ExpectToHaveFinalizer(trainingJob, controllers.SageMakerResourceFinalizerName)
+					})
+				})
+			})
+
+			When("HasDeletionTimestamp", func() {
+				BeforeEach(func() {
+					shouldHaveDeletionTimestamp = true
+				})
+
+				It("Deletes the training job", func() {
+					ExpectTrainingJobToBeDeleted(trainingJob)
```

Comment:
`Is this right? When the training job status is "Failed" and the debug job status is "Stopping", we will allow the Kubernetes job to be deleted? `
There is not k8s job for debug job, so we are not deleting anything. The debug job can be in stopping state due to various reasons. 

`Do the debug jobs charge the user while they run? If so I think we should maybe wait until the DebugJobs are finished before we delete the etcd entry. Whatever we do here, we should do for all other HasDeletionTimestamp debug job cases.`

I don't know about pricing, couldn't find any reference. I don't think we should do that way, ideally debugger should have been designed to be like a spawner job (similar to HPO). With current state I am a bit hesitant to go in that direction.

Reply:
I agree with @gautamkmr here in that our job was to manage the training job that they started, and if they attach a debugger then they assume any costs incurred to that configuration. If the user sees their TrainingJob as "Failed" then they would assume it would be possible to delete it - it may be unclear as to why we are restricting this ability.